autoarray/structures/grids/grid_2d_util.py

```diff
@@ -250,8 +282,12 @@ def grid_2d_slim_via_mask_from(
     for y in range(mask_2d.shape[0]):
         for x in range(mask_2d.shape[1]):
             if not mask_2d[y, x]:
-                grid_slim[index, 0] = -(y - centres_scaled[0]) * pixel_scales[0]
-                grid_slim[index, 1] = (x - centres_scaled[1]) * pixel_scales[1]
+                if use_jax:
+                    grid_slim = grid_slim.at[index, 0].set(-(y - centres_scaled[0]) * pixel_scales[0])
+                    grid_slim = grid_slim.at[index, 1].set((x - centres_scaled[1]) * pixel_scales[1])
+                else:
+                    grid_slim[index, 0] = -(y - centres_scaled[0]) * pixel_scales[0]
+                    grid_slim[index, 1] = (x - centres_scaled[1]) * pixel_scales[1]

```

Comment:
It's funny that JAX can't use `__getitem__` and `__setitem__` to mimic the numpy interface here

Reply:
It has more to do with how JAX can't update arrays in place, it must return a new object.  This has to do with ensuring that the compute graph is "one direction" and doesn't have any closed loops (this ensures that the graph can be auto-differentiated).

Note: once the code is `jit`ed it will turn this back into an in-place update.  It is just the initial pre-optimized XLA compute graph that needs to be non-looping.  This is also why classes that differentiate variables on `self` need to define a flattening and un-flattening function that makes a new class instance each time a variable is updated.

Reply:
Oh that's interesting. I guess it would be very inefficient to be allocating a new array every time. Not sure I see how assignment could cause a closed-loop in the compute graph though?